plugins/meta/bandwidth/ifb_creator.go

```diff
@@ -105,43 +116,198 @@ func CreateEgressQdisc(rateInBits, burstInBits uint64, hostDeviceName string, if
 	}
+
+	// Step 2 classes
+
 	rateInBytes := rateInBits / 8
 	burstInBytes := burstInBits / 8
 	bufferInBytes := buffer(rateInBytes, uint32(burstInBytes))
-	latency := latencyInUsec(latencyInMillis)
-	limitInBytes := limit(rateInBytes, latency, uint32(burstInBytes))
 
-	qdisc := &netlink.Tbf{
-		QdiscAttrs: netlink.QdiscAttrs{
+	// The capped class for all but excluded subnets
+	// cmd = exec.Command("/usr/sbin/tc", "class", "add", "dev", interfaceName, "parent", "1:", "classid", "1:30", "htb", "rate",
+	//       fmt.Sprintf("%d", rateInBits), "burst", fmt.Sprintf("%d", burstInBits))
+	defClass := &netlink.HtbClass{
+		ClassAttrs: netlink.ClassAttrs{
 			LinkIndex: linkIndex,
-			Handle:    netlink.MakeHandle(1, 0),
-			Parent:    netlink.HANDLE_ROOT,
+			Handle:    netlink.MakeHandle(1, DefaultClassMinorID),
+			Parent:    netlink.MakeHandle(1, 0),
 		},
-		Limit:  limitInBytes,
 		Rate:   rateInBytes,
 		Buffer: bufferInBytes,
+		// Let's set up the "burst" rate to twice the specified rate
+		Ceil:    2 * rateInBytes,
+		Cbuffer: bufferInBytes,
 	}
-	err := netlink.QdiscAdd(qdisc)
+
+	err = netlink.ClassAdd(defClass)
+	if err != nil {
+		return fmt.Errorf("error while creating htb default class: %s", err)
+	}
+
+	// The uncapped class for the excluded subnets
+	// cmd = exec.Command("/usr/sbin/tc", "class", "add", "dev", interfaceName, "parent", "1:", "classid", "1:1", "htb",
+	// 	"rate", "100000000000")
+	bigRate := UncappedRate
+	uncappedClass := &netlink.HtbClass{
+		ClassAttrs: netlink.ClassAttrs{
+			LinkIndex: linkIndex,
+			Handle:    netlink.MakeHandle(1, 1),
+			Parent:    qdisc.Handle,
+		},
+		Rate: bigRate,
+		Ceil: bigRate,
+		// No need for any burst, the minimum buffer size in q_htb.c should be enough to handle the rate which
+		// is already more than enough
+	}
+	err = netlink.ClassAdd(uncappedClass)
 	if err != nil {
-		return fmt.Errorf("create qdisc: %s", err)
+		return fmt.Errorf("error while creating htb uncapped class: %s", err)
+	}
+
+	// Now add filters to redirect excluded subnets to the class 1 instead of the default one (30)
+
+	for _, subnet := range excludeSubnets {
+
+		// cmd = exec.Command("/usr/sbin/tc", "filter", "add", "dev", interfaceName, "parent", "1:", "protocol", protocol,
+		// "prio", "16", "u32", "match", "ip", "dst", subnet, "flowid", "1:1")
+
+		_, nw, err := net.ParseCIDR(subnet)
+		if err != nil {
+			return fmt.Errorf("bad subnet %s: %s", subnet, err)
+		}
+		var maskBytes []byte = nw.Mask
+		var subnetBytes []byte = nw.IP
+
+		if len(maskBytes) != len(subnetBytes) {
+			return fmt.Errorf("error using net lib for subnet %s len(maskBytes) != len(subnetBytes) "+
+				"(%d != %d) should not happen", subnet, len(maskBytes), len(subnetBytes))
+		}
+
+		isIpv4 := nw.IP.To4() != nil
+		// protocol := syscall.ETH_P_IPV6
+		var offset int32 = 24
+		keepBytes := 16
+		if isIpv4 {
+			// protocol = syscall.ETH_P_IP
+			offset = 16
+			keepBytes = 4
+
+		}
+
+		protocol := syscall.ETH_P_ALL
```

Comment:
why `protocol` is all, instead of `ip' or `ipv6' (as you commented just above?)

Reply:
fixed